torch/_functorch/_aot_autograd/collect_metadata_analysis.py

```diff
@@ -91,16 +99,62 @@ def coerce_tangent(x):
         out, "__coerce_tangent_metadata__"
     ):
         out = out.__coerce_tangent_metadata__()  # type: ignore[attr-defined]
-    # It's possible to have a subclass that advertises as contiguous,
-    # but has noncontiguous inner tensors.
-    # Force these to be conntiguous too
+
+    # Coerce subclasses elements to memory_format too.
     if is_traceable_wrapper_subclass(out):
-        for attr in out.__tensor_flatten__()[0]:  # type: ignore[attr-defined]
+        attrs = out.__tensor_flatten__()[0]
+
+        assert isinstance(memory_format, list)
+        for i, attr in enumerate(attrs):
+            elem = getattr(out, attr)
+            new_elem, updated = coerce_tangent(elem, memory_format[1 + i])
+            if updated:
+                setattr(out, attr, new_elem)
+
+    return out, updated
+
+
+# Coercing and collecting traced tangents memory format in one recursive traversal
+# mypy: ignore-errors
+def coerce_tangent_and_suggest_memory_format(
+    x: Tensor, force_memory_format: Optional[torch.memory_format] = None
```

Comment:
hmm... do we actually pass in a non-None value for `force_memory_format` anywhere? I don't think I see one.

Given that we have a separate coercion function for trace-time vs runtime, the "force_memory_format" seems unnecessary? (the trace time fn never forces, the runtime fn always forces)

Reply:
Yes, non-None value means that we need to run suggest_memory_formt to deduce it, this happens for all output tangents. While for others - aliases, mutation input tangents - it will have a value.